request/request.go

```diff
@@ -276,16 +276,21 @@ func (r *Request) Name() string {
 	if r.name != "" {
 		return r.name
 	}
-	if r.Req == nil {
-		r.name = "."
-		return "."
-	}
-	if len(r.Req.Question) == 0 {
-		r.name = "."
-		return "."
+
+	r.name = "."
+	switch {
+	case r.Req == nil:
+	case len(r.Req.Question) > 0:
+		r.name = strings.ToLower(r.Req.Question[0].Name)
+	case len(r.Req.Stateful) > 0:
+		switch tlv := r.Req.Stateful[0].(type) {
```

Comment:
This relies on API that does not exist (yet) in the dns package. But it's being worked on and one way or another these TLVs will be exposed.